### ISSUE TITLE
Map not updating on certain Xiaomi devices using Android Auto

### ISSUE THREAD

**Issue by g4rb4g3** [mapbox/mapbox-maps-android#1413]

**Comment by kmadsen** [comment]
Fixed an issue in the examples repo, there is still a issue that may be resolved in the Maps repository specific to Xiaomi devices

**Comment by g4rb4g3** [comment]
@kmadsen Thank you for pointing me to the other sample app I like it a lot since it doesn't work at all on the Xiaomi device. 😉 It doesn't have any activity so if you just launch the app on Android Auto it is stuck on the position it got while starting up.
I added some logging and you can see that the new position is incoming. I also added 2 listeners one for renderFrameStarted and another one for renderFrameFinished. Additionally I added an activity I can launch on the phone to make the map updates work. You can check out all my changes to the app [here](https://github.com/mapbox/mapbox-maps-android/compare/main...g4rb4g3:main).

So from the [logs](https://gist.github.com/g4rb4g3/f7af85901366d0b4e02bfbcdb450ea71) you can see that there are quite some call for addOnRenderFrameStartedListener and addOnRenderFrameFinishedListener in the beginning (app started on Android Auto only).
Once everything settled you can see many of these: `2022-06-07 10:44:25.374 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  changePositionListener, got new position Point{type=Point, bbox=null, coordinates=[14.0318107, 48.2240837]}
2022-06-07 10:44:25.375 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  changeBearingListener, got new bearing 0.0
2022-06-07 10:44:25.438 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameStartedListener
2022-06-07 10:44:25.440 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  changePositionListener, got new position Point{type=Point, bbox=null, coordinates=[14.031682699600694, 48.224026310360316]}
2022-06-07 10:44:25.442 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  changeBearingListener, got new bearing -7.598346638515068
2022-06-07 10:44:25.458 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameFinishedListener
2022-06-07 10:44:25.461 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  changePositionListener, got new position Point{type=Point, bbox=null, coordinates=[14.031651669195607, 48.224012397718035]}
2022-06-07 10:44:25.465 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  changeBearingListener, got new bearing -9.440370377946238
2022-06-07 10:44:25.475 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameStartedListener`
So a new position is arriving but addOnRenderFrameStartedListener and addOnRenderFrameFinishedListener are called once per gps update mostly.

Only when I launch the activity on the phone you can see that the listeners are reporting like we would expect them to and map updates start working on Android Auto.
`2022-06-07 10:44:30.381 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  changePositionListener, got new position Point{type=Point, bbox=null, coordinates=[14.02916881403077, 48.22313784773048]}
2022-06-07 10:44:30.381 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  changeBearingListener, got new bearing 237.4023516861871
2022-06-07 10:44:30.389 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameFinishedListener
2022-06-07 10:44:30.402 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameStartedListener
2022-06-07 10:44:30.406 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameFinishedListener
2022-06-07 10:44:30.416 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameStartedListener
2022-06-07 10:44:30.422 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameFinishedListener
2022-06-07 10:44:30.432 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameStartedListener
2022-06-07 10:44:30.437 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameFinishedListener
2022-06-07 10:44:30.450 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameStartedListener
2022-06-07 10:44:30.453 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameFinishedListener
2022-06-07 10:44:30.465 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameStartedListener
2022-06-07 10:44:30.469 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameFinishedListener
2022-06-07 10:44:30.473 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameStartedListener
2022-06-07 10:44:30.477 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameFinishedListener
2022-06-07 10:44:30.481 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameStartedListener
2022-06-07 10:44:30.486 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameFinishedListener
2022-06-07 10:44:30.497 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameStartedListener
2022-06-07 10:44:30.505 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameFinishedListener
2022-06-07 10:44:30.514 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameStartedListener
2022-06-07 10:44:30.517 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameFinishedListener
2022-06-07 10:44:30.522 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameStartedListener
2022-06-07 10:44:30.526 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameFinishedListener
2022-06-07 10:44:30.530 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameStartedListener
2022-06-07 10:44:30.534 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameFinishedListener
2022-06-07 10:44:30.539 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameStartedListener
2022-06-07 10:44:30.542 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameFinishedListener
2022-06-07 10:44:30.556 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameStartedListener
2022-06-07 10:44:30.556 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameFinishedListener
2022-06-07 10:44:30.566 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameStartedListener
2022-06-07 10:44:30.571 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  addOnRenderFrameFinishedListener
2022-06-07 10:44:30.576 10388-10388 CarCameraController     com.mapbox.maps.testapp.auto         D  changePositionListener, got new position Point{type=Point, bbox=null, coordinates=[14.02916881403077, 48.22313784773048]}`

**Comment by 2flea** [comment]
Is there any update on this issue? Any fix in sight? I use same phone as @g4rb4g3 - Poco F3 with Android 12 (latest update). I want to use ABRP with AndroidAuto but I need to have the app open and active on my phone all the time. Cheers,
Andreas

**Comment by SamuelBrucksch** [comment]
@kmadsen could you please give an update on the process of this? We are getting more and more reports from Xiaomi users. As this seems to be reproducible easily, could you please rank this a but higher so we can release an update for the affected users before the summer holidays start?

**Comment by SamuelBrucksch** [comment]
@pengdev @tobrun @kmadsen can anyone please comment on this? This should really affect the usage of this library for a lot of users, if it can't work on Xiaomi unless the phone screen is on...

**Comment by pengdev** [comment]
Hi @SamuelBrucksch thanks for reporting the issue, we are still trying to get access to a compatible Xiaomi device to reproduce this issue, will keep you posted about the progress.

**Comment by CarstenWa** [comment]
It isn't just xiaomi. Have the same issues on my huawei p30 pro

**Comment by xflow7** [comment]
Hi, @pengdev and @kmadsen, for what it's worth I'm also not sure how device-specific this is. I have the same problem on a Samsung S21. (I arrived here from [this bug thread on the ABRP forum](https://abrp.upvoty.com/b/report-a-bug/map-doesnt-update-on-android-auto-when-phone-display-is-off/))

**Comment by SamuelBrucksch** [comment]
More and more reports incoming on this. We had to buy a Poco F3 ourself to replicate the problem. If you can't find anyone with a phone please order one so you can fix this.

**Comment by SamuelBrucksch** [comment]
> Hi, @pengdev and @kmadsen, for what it's worth I'm also not sure how device-specific this is. I have the same problem on a Samsung S21. (I arrived here from [this bug thread on the ABRP forum](https://abrp.upvoty.com/b/report-a-bug/map-doesnt-update-on-android-auto-when-phone-display-is-off/))

Samsung is a separate issue and not related to this. A fix will be available soon.

We got another report from Xiaomi Mi11 pro. So it seems like any Xiaomi device on Android 12 will do.

**Comment by CarstenWa** [comment]
Got the same Problem on an older Huawai p20 with Android 10 on it

**Comment by kmadsen** [comment]
I've ordered a Poco F3 but it'll be here tomorrow at some point

**Comment by CarstenWa** [comment]
> I've ordered a Poco F3 but it'll be here tomorrow at some point

If it helps out with Huawai I will donate a bit 😁

**Comment by kmadsen** [comment]
I have a Xiaomi Poco F3. I was not able to reproduce on the out of the box version - on Android 11. I also tested with multiple mock GPS location apps ([GPS Locations](https://play.google.com/store/apps/details?id=ru.gavrikov.mocklocations) and [Mock GPS with joystick](https://play.google.com/store/apps/details?id=net.marlove.mockgps)). 

> Poco F3 with Android 12 (latest update)

I'm updating the phone to test the latest update with Android 12, but posting this because it does work on certain Xiaomi builds.

https://user-images.githubusercontent.com/3021882/180613960-63fdbbe6-0917-4b0f-8a5a-a700da347d0e.mov

**Comment by g4rb4g3** [comment]
@kmadsen Thanks for the update, as pointed out in the first post the issue happens on Android 12 for sure. I did not test it with Android 11 since I did the update straight after unpacking it.

**Comment by kmadsen** [comment]
The system updates available are for the MIUI, so I updated to MIUI 12.5.6 and it worked again.

I'm not seeing an option to upgrade to Android 12, but am seeing an option to upgrade the MIUI (next is 13.0.4.0)

**Comment by SamuelBrucksch** [comment]
Yes MIUI 13 is Android 12. People told us this issue started after the update, so according to them it stilled worked on Android 11.

**Comment by kmadsen** [comment]
Ok sounds good, thanks! I'm about to go on a hike (test drive to the trailhead), so after the update I'll be able to see the issues on a few apps. Will dig into potential solutions when I'm back by a computer.

**Comment by kmadsen** [comment]
It's still working for me after the update, on the desktop head unit. I'm getting the same result as the video here https://github.com/mapbox/mapbox-maps-android/issues/1413#issuecomment-1193151608

Model: [M2012K11AG](https://www.gsmarena.com/xiaomi_poco_f3-10758.php)
MIUI version: 13.0.4Stable, 13.0.4.0(SKHMIXM)
Android version: 12 SKQ1.211006.001
Android security update 2022-06-01

I'm still going for a test drive to see the non-mocked locations. But is there potentially, another setting that needs to be enabled/disabled in order to reproduce this? For example, the [GPS Locations](https://play.google.com/store/apps/details?id=ru.gavrikov.mocklocations) suggests this and I haven't enabled that yet.

![Screen Shot 2022-07-23 at 10 17 13 AM](https://user-images.githubusercontent.com/3021882/180615801-541f0684-5ac0-40e6-b17e-3338594f13d4.png)

**Comment by CarstenWa** [comment]
On Android 10 it won't work also.

**Comment by SamuelBrucksch** [comment]
We only used the sample apps as they are and didn't do anything else. The phone screen must be off for the not moving map to show.

**Comment by kmadsen** [comment]
OK I am able to reproduce the OP issue locally. I'm unable to reproduce in the [android-auto-app](https://github.com/mapbox/mapbox-maps-android/tree/main/android-auto-app) as mentioned here https://github.com/mapbox/mapbox-maps-android/issues/1413#issuecomment-1148398491. My repo steps require you to first, have an open mobile app and then close it.

## Devices reproduced on

Xiaomi MIUI 13.0.4
Android version 12 SKQ1.211006.001

## Repo steps

1. Launch an app with an Activity ([qa-test-app](https://github.com/mapbox/mapbox-navigation-android/tree/main/qa-test-app) or [android-auto-app](https://github.com/mapbox/mapbox-navigation-android-examples/tree/main/android-auto-app))
2. Launch the desktop head unit (for example: [$ make car](https://github.com/mapbox/mapbox-navigation-android/blob/73f0ea7e2177d41eaad13ec4f80f24199a39146d/Makefile#L223-L225))
3. Kill the mobile app (or put app in background)
4. Animations have stopped working
5. Scroll the map
6. Try to recenter the map, notice the animation will not work

## Looks like an animation issue

I am able to fix the recenter button by [removing the usage of AnimatorSet](https://github.com/mapbox/mapbox-navigation-android/blob/73f0ea7e2177d41eaad13ec4f80f24199a39146d/libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/camera/transition/MapboxNavigationCameraTransition.kt#L40-L44)
``` kotlin
        mapboxMap.setCamera(cameraOptions)
        return AnimatorSet()
//        return ifNonNull(pluginImpl) {
//
//            flyFromLowZoomToHighZoom(cameraOptions, it, transitionOptions)
//        } ?: fromLowZoomToHighZoom(cameraOptions, transitionOptions)
```

## Video and logs

I added a log animator to the [MapboxNavigationCameraTransition](https://github.com/mapbox/mapbox-navigation-android/blob/73f0ea7e2177d41eaad13ec4f80f24199a39146d/libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/camera/transition/MapboxNavigationCameraTransition.kt#L189). The logs include an expected update (animates a value from 0 to 10). When the activity is backgrounded, the value animator will still be triggered but it animates from 0 to 0. Resulting in no movement from the animations.

https://user-images.githubusercontent.com/3021882/180879867-15a3c473-a308-48f0-bb7c-f437ed5cee06.mov

``` kotlin
    inner class LogAnimator : ValueAnimator() {
        init {
            addPauseListener(
                onResume = { logI("onResume", "kyle_debug") },
                onPause = { logI("onPause", "kyle_debug") },
            )
            addListener(
                onStart = { logI("onStart", "kyle_debug") },
                onCancel = { logI("onCancel", "kyle_debug") },
                onEnd = { logI("onEnd", "kyle_debug") },
                onRepeat = { logI("onRepeat", "kyle_debug") },
            )
            addUpdateListener {
                logI("onUpdate ${it.animatedValue as Int}", "kyle_debug")
            }
            setIntValues(0, 10)
        }
    }
```

```
2022-07-25 14:57:04.663 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onStart
2022-07-25 14:57:04.663 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 0
2022-07-25 14:57:04.663 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 0
2022-07-25 14:57:04.674 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 0
2022-07-25 14:57:04.688 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 0
2022-07-25 14:57:04.702 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 0
2022-07-25 14:57:04.719 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 0
2022-07-25 14:57:04.736 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 1
2022-07-25 14:57:04.752 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 1
2022-07-25 14:57:04.770 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 2
2022-07-25 14:57:04.786 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 3
2022-07-25 14:57:04.803 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 4
2022-07-25 14:57:04.819 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 4
2022-07-25 14:57:04.837 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 5
2022-07-25 14:57:04.854 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 6
2022-07-25 14:57:04.870 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 7
2022-07-25 14:57:04.886 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 8
2022-07-25 14:57:04.903 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 8
2022-07-25 14:57:04.920 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 9
2022-07-25 14:57:04.934 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 9
2022-07-25 14:57:04.950 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 9
2022-07-25 14:57:04.967 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 9
2022-07-25 14:57:04.983 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 10
2022-07-25 14:57:04.983 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onEnd
2022-07-25 14:57:05.686 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onStart
2022-07-25 14:57:05.686 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 0
2022-07-25 14:57:05.686 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 0
2022-07-25 14:57:06.659 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onCancel
2022-07-25 14:57:06.659 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onEnd
2022-07-25 14:57:06.660 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onStart
2022-07-25 14:57:06.660 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 0
2022-07-25 14:57:06.660 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 0
2022-07-25 14:57:07.658 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onCancel
2022-07-25 14:57:07.658 6528-6528/com.mapbox.navigation.qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onEnd
```

**Comment by g4rb4g3** [comment]
I just updated my phone to the latest firmware:
MIUI: MIUI Global 13.0.7 Stable (SKHEUXM)
Android 12 SKQ1.211006.001

Still seeing the same behavior like in the first post.
I added some logging to MapboxNavigationCameraTransition.kt and I see the same, as soon as the screen is off or the app is not in foreground the centerAnimator gets the same coordinates over and over. 
https://user-images.githubusercontent.com/646435/180968316-9ffaeec4-9090-43d6-997c-5218f52f6900.mp4


To create some nice screen captures you can use [scrcpy](https://github.com/Genymobile/scrcpy)

**Comment by kmadsen** [comment]
Good idea with scrcpy :). Next, wanted to find if something in Mapbox SDKs is canceling the animations. Here is an empty project, reproducing the issue without a Mapbox SDK [hellocar.zip](https://github.com/mapbox/mapbox-maps-android/files/9191021/hellocar.zip)

This is most likely an issue with the `MIUI` system disabling animations. I am still unable to reproduce on Pixel 4XL which is running Android 12.

https://user-images.githubusercontent.com/3021882/181044841-4c1883fe-9561-4c0f-b2fd-e65465141722.mov

``` kotlin
class MainSession : Session() {
--snip--
    init {
        lifecycle.addObserver(object : DefaultLifecycleObserver {
            override fun onCreate(owner: LifecycleOwner) {
                var lastAnimator: ValueAnimator? = null
                lifecycle.coroutineScope.launch {
                    while (isActive) {
                        lastAnimator?.cancel()
                        lastAnimator = LogAnimator().apply { start() }
                        delay(1000)
                    }
                }
            }
        })
    }
}
```

**Comment by kmadsen** [comment]
I've submitted this feedback to xiaomi because this appears to be specific to `MIUI` https://xiaomiui.net/how-to-report-bugs-on-miui-7934/. I'm not sure of a way to reference open issues for xiaomi. 

![Screenshot_2022-07-26-08-31-53-681_com miui bugreport](https://user-images.githubusercontent.com/3021882/181049284-4056f9f2-c049-47a3-ac77-e330eab034d2.jpg)

@g4rb4g3 @CarstenWa @xflow7 @SamuelBrucksch you've mentioned this issue is reproducible on other devices. Can you reproduce the issue with the [hellocar.zip](https://github.com/mapbox/mapbox-maps-android/files/9191021/hellocar.zip) project from this https://github.com/mapbox/mapbox-maps-android/issues/1413#issuecomment-1195627554?

I would like to know if which devices contain this issue or if other devices have a different issue.

**Comment by kmadsen** [comment]
Another thing tested, is that there exists an Android 12 setting to disable animations. https://mcmw.abilitynet.org.uk/how-to-disable-interface-animations-in-android-12

On Xiaomi, when the animations are paused it emits the start value multiple times https://github.com/mapbox/mapbox-maps-android/issues/1413#issuecomment-1194685226. This would be at least better, if the update emitted the final value (instead of the first value twice). 
```
2022-07-25 15:08:05.055 qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onStart
2022-07-25 15:08:05.055 qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 5.0
2022-07-25 15:08:05.055 qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onUpdate 5.0
2022-07-25 15:08:06.043 qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onCancel
2022-07-25 15:08:06.044 qa_test_app I/Mapbox: [nav-sdk]: [kyle_debug] onEnd
```

Here is what happens when disabling animations on the Xiaomi device and then backgrounding the app. Notice, disabling animations works as expected (first and last value). But when backgrounding the app it only emits the first value. So this is difficult behavior that Xiaomi devices create with animations.
``` 
2022-07-26 08:59:59.918 13259-13259/com.mapbox.example.hellocar I/kyle_debug: onEnd
2022-07-26 09:00:00.909 13259-13259/com.mapbox.example.hellocar I/kyle_debug: onStart
2022-07-26 09:00:00.910 13259-13259/com.mapbox.example.hellocar I/kyle_debug: onUpdate 5.0
2022-07-26 09:00:00.923 13259-13259/com.mapbox.example.hellocar I/kyle_debug: onUpdate 10.0
2022-07-26 09:00:00.927 13259-13259/com.mapbox.example.hellocar I/kyle_debug: onEnd
2022-07-26 09:00:01.912 13259-13259/com.mapbox.example.hellocar I/kyle_debug: onStart
2022-07-26 09:00:01.913 13259-13259/com.mapbox.example.hellocar I/kyle_debug: onUpdate 5.0
2022-07-26 09:00:02.916 13259-13259/com.mapbox.example.hellocar I/kyle_debug: onCancel
2022-07-26 09:00:02.916 13259-13259/com.mapbox.example.hellocar I/kyle_debug: onEnd
2022-07-26 09:00:02.917 13259-13259/com.mapbox.example.hellocar I/kyle_debug: onStart
```

**Comment by SamuelBrucksch** [comment]
Hi @kmadsen 

great analysis. However i would not expect Xiaomi to fix this in a timely manner. TomTom for example does not seem to rely on Animators. Could you please check if there are other way on how to fix or at least workaround this?

I don't have a Xiaomi device myself, so I unfortunately can't test it.

**Comment by kmadsen** [comment]
> great analysis. However i would not expect Xiaomi to fix this in a timely manner. Could you please check if there are other ways on how to fix or at least workaround this?

Am investigating possible workarounds. The `ValueAnimator` uses the [Choreographer](https://developer.android.com/reference/android/view/Choreographer), and the Mapbox SDK relies on the `Choreographer` for not just animations, but also [map rendering](https://github.com/mapbox/mapbox-maps-android/blob/ce278875b87db4f7ff66594ddf866f0ddf5da3e7/sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt#L31). 

I'm finding that, Xiaomi will not post calls made to `Choreographer.getInstance().postFrameCallback` when the app is backgrounded. When I'm attempting to debug the source of `Choreographer` on a Xiaomi device, the source code is not the same as a Pixel on the same Android version. MIUI has a custom implementation, and it is possible that the Xiaomi team has not been optimizing for Android Auto much yet. If Android Auto gets to be a super popular framework, how would Xiaomi support animations with this behavior 🤔 ?

A Mapbox solution seems to require a migration off of the `Choreographer` and that would take careful planning. Building a new animation and rendering framework for AA is also, not a quick fix.

**Comment by yunikkk** [comment]
> I'm finding that, Xiaomi will not post calls made to `Choreographer.getInstance().postFrameCallback` when the app is backgrounded.

@kmadsen so map rendering is fully broken and not only animations, right? Even manual camera updates won't work if Choreographer is paused/disposed.
Assume Xiaomi just hadn't count on Android Auto usecases and should fix this on their end. Wonder if there are tricky hacks to reenable Choreographer...

**Comment by kmadsen** [comment]
> so map rendering is fully broken and not only animations, right?

@yunikkk Wait.. the render thread is still working. The reason is because the [MapboxRenderThread](https://github.com/mapbox/mapbox-maps-android/blob/ce278875b87db4f7ff66594ddf866f0ddf5da3e7/sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt#L471) is using a background thread. This will work for animations!

Here is an animator that will continue to work even when the app is backgrounded. The Xiaomi device is killing **main thread workers**. Nice, this is a manageable work-around

https://user-images.githubusercontent.com/3021882/182428402-eb95b093-bcef-4274-962c-aae65fe29853.mov

``` kotlin
class LogAnimation(
    val duration: Long
): Choreographer.FrameCallback {
    private var startTime = 0L
    private var isRunning = false
    private var handlerThread: HandlerThread? = null
    private var handler: Handler? = null

    fun start() {
        isRunning = true
        startTime = System.nanoTime()
        handlerThread = HandlerThread(HANDLE_THREAD_NAME, THREAD_PRIORITY_DISPLAY).apply {
            start()
            handler = Handler(this.looper)
        }
        post {
            Choreographer.getInstance().postFrameCallback(this)
        }
        Log.i("animation_debug", "LogAnimation onStart")
    }

    private fun post(task: () -> Unit) {
        handler?.let {
            val message = Message.obtain(it, task)
            message.obj = EVENT
            it.sendMessageDelayed(message, 0)
        } ?: logW("animation_debug", "Animation thread was not started, ignoring event")
    }

    fun cancel() {
        handler?.removeCallbacksAndMessages(EVENT)
        handlerThread?.quitSafely()
        handler = null
        if (isRunning) {
            Log.i("animation_debug", "LogAnimation onCancel")
        }
        isRunning = false
    }

    @WorkerThread
    override fun doFrame(frameTimeNanos: Long) {
        val timeDiff = (frameTimeNanos - startTime) / 1000000.0
        if (isRunning) {
            if (timeDiff >= duration) {
                Log.i("animation_debug", "LogAnimation onCompleted $timeDiff")
                isRunning = false
            } else {
                Choreographer.getInstance().postFrameCallback(this@LogAnimation)
                Log.i("animation_debug", "LogAnimation onUpdate $timeDiff")
            }
        }
    }

    companion object {
        private const val HANDLE_THREAD_NAME = "animation_thread"
        private val EVENT: Any = Object()
    }
}
```

**Comment by SamuelBrucksch** [comment]
Great progress. Let me know once there is a snapshot or beta version with the workaround implemented, so we can try it in our app.

**Comment by rychuelektryk** [comment]
Hi, great investigation. Any news?

**Comment by kmadsen** [comment]
This will take some time. It is an odd behavior from Xiaomi to require animations to be started from a background thread.

There are some [assumptions](https://github.com/mapbox/mapbox-maps-android/issues/1556) that animations are started from the main thread, and we use android animator framework for location and camera animations.

**Comment by SamuelBrucksch** [comment]
Hi @kmadsen

can you please provide a rough estimate when this will be fixed? We are losing more and more users due to this.

**Comment by kiryldz** [comment]
> Hi @kmadsen
> 
> can you please provide a rough estimate when this will be fixed? We are losing more and more users due to this.

@SamuelBrucksch I've drafted implementation to allow using background thread for animations [here](https://github.com/mapbox/mapbox-maps-android/pull/1638). Next step is to incorporate that into AA + Nav SDK if needed. 

Rough optimistic estimate: v10.9.0-beta.1 Maps release, happening in 3 weeks.

**Comment by SamuelBrucksch** [comment]
Ok that sounds great. Please let me know as soon as there is a snapshot version we can use for testing. We only use Mapbox and not the AA / Nav SDK, so once the change is in mapbox we can already test this.

**Comment by kiryldz** [comment]
@SamuelBrucksch you could try `10.8.0-beta.1-MAPSAND-395-introduce-runtime-option-to-run-animators-on-non-main-thread-2dcb37457-SNAPSHOT`, just add additional mirror as described in https://github.com/mapbox/mapbox-maps-android/blob/main/DEVELOPING.md#working-with-snapshots.

**Comment by SamuelBrucksch** [comment]
I will, is there anything we have to do in our code too? In the PR i saw that you can now switch between background and main runner. Is that done automatically or do I have to switch myself for certain devices?

**Comment by kiryldz** [comment]
> I will, is there anything we have to do in our code too? In the PR i saw that you can now switch between background and main runner. Is that done automatically or do I have to switch myself for certain devices?

That snapshot provided above does not yet have AA related code that @kmadsen added so everything should be enabled from your end for now.

**Comment by SamuelBrucksch** [comment]
Ok so if we notice, that we have some problematic phones like Xiaomi on Android 12+, i just have to call `AnimationThreadController.useBackgroundThread()` before initializing mapbox? Or is there anything else i have to do?

**Comment by SamuelBrucksch** [comment]
No matter what i try, i always end up with this exception:
```
2022-09-02 10:25:52.380 14090-14181/? E/AndroidRuntime: FATAL EXCEPTION: MapboxAnimThread
    Process: com.iternio.abrpapp, PID: 14090
    java.lang.RuntimeException: Function 'postOnAnimatorThread' could not be used from non-main thread when 'useBackgroundThread()' was not called beforehand!
        at com.mapbox.maps.threading.AnimationThreadController.postOnAnimatorThread(AnimationThreadController.kt:89)
        at com.mapbox.maps.plugin.animation.animator.CameraAnimator.addListener(CameraAnimator.kt:102)
        at android.animation.AnimatorSet.addAnimationEndListener(AnimatorSet.java:1248)
        at android.animation.AnimatorSet.startAnimation(AnimatorSet.java:1189)
        at android.animation.AnimatorSet.start(AnimatorSet.java:729)
        at android.animation.AnimatorSet.start(AnimatorSet.java:684)
        at com.mapbox.maps.plugin.animation.CameraAnimationsPluginImpl$startHighLevelAnimation$2$1.invoke(CameraAnimationsPluginImpl.kt:903)
        at com.mapbox.maps.plugin.animation.CameraAnimationsPluginImpl$startHighLevelAnimation$2$1.invoke(CameraAnimationsPluginImpl.kt:903)
        at com.mapbox.maps.threading.AnimationThreadController.postOnAnimatorThread$lambda-2(AnimationThreadController.kt:85)
        at com.mapbox.maps.threading.AnimationThreadController.$r8$lambda$nE-XMEhVlqLA1g9jOyl1_QBh_Io(Unknown Source:0)
        at com.mapbox.maps.threading.AnimationThreadController$$ExternalSyntheticLambda0.run(Unknown Source:2)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loopOnce(Looper.java:201)
        at android.os.Looper.loop(Looper.java:288)
        at android.os.HandlerThread.run(HandlerThread.java:67)
```

I call `AnimationThreadController.useBackgroundThread()` in `onSurfaceAvailable()` before we initialize mapbox, tried it at the end when all is initialized before we actually call any animations but it ended up in the same result. Any advice @kiryldz ?

**Comment by kiryldz** [comment]
@SamuelBrucksch - if you have some preconditions like (AA + Xiaomi + Android 12+) - then better better enable that for this use-case _only_. Additionally I'd advise to run `useBackgroundThread()` only when going to background on device (locking the screen) and use `useMainThread()` everywhere else.

About the exception - I did publish latest snapshot `10.8.0-beta.1-MAPSAND-395-introduce-runtime-option-to-run-animators-on-non-main-thread-6d5eab9d4-SNAPSHOT` based on [this](https://github.com/mapbox/mapbox-maps-android/pull/1638) branch, hopefully it should fix it. Also you could take a look how Kyle implemented enabling background animation thread for AA case only when going to background and then enabling main animation thread again when returning to the foreground.

**Comment by SamuelBrucksch** [comment]
Thanks @kiryldz it seems to work now, tried it on a non Xiaomi phone with background always on. However it seems to have some hickups in the animator. Please check the video below

https://user-images.githubusercontent.com/8776268/188110647-72d25b05-4390-478b-b56d-c848664b5ed8.mp4

Is that something than can also be fixed easily? Because those jumps are quite distracting.

**Comment by kiryldz** [comment]
> Thanks @kiryldz it seems to work now, tried it on a non Xiaomi phone with background always on. However it seems to have some hickups in the animator. Please check the video below
> 
>  PXL_20220902_093231282.mp4 
> Is that something than can also be fixed easily? Because those jumps are quite distracting.

Well, it could be tonns of things... 
1. Are you using `ViewportPlugin`?
2. Could you please enable `CameraAnimationsPlugin.debugMode` and take a look at the logs?

**Comment by SamuelBrucksch** [comment]
This is definitely related to the background animators, as it works when disabling the background thread. We don't use the ViewportPlugin, just plain camera animation.

However we call animator.cancel() before starting the next animation so that two animations that follow quickly after another do not cause any strange side effects if the first did not finish when the second starts.

```
public void animateMapTo(Point mapCenter, double bearing, double pitch, double zoom) {
        if (mapAnimator != null) {
            mapAnimator.cancel();
            mapAnimator = null;
        }

        mapAnimator = cameraAnimations.easeTo(currentCamera.build(), new MapAnimationOptions.Builder().duration(1000).interpolator(linearInterpolator).build());
    }
```

Debug mode shows this:
```
2022-09-02 12:00:22.744 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation CENTER was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:22.744 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation BEARING was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:22.744 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation PADDING was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:22.744 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation PITCH was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:22.744 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation ZOOM was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:23.748 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation CENTER was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:23.748 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation BEARING was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:23.748 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation PADDING was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:23.748 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation PITCH was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:23.748 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation ZOOM was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:24.752 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation CENTER was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:24.752 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation BEARING was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:24.752 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation PADDING was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:24.752 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation PITCH was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:24.752 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation ZOOM was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:25.755 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation CENTER was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:25.756 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation BEARING was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:25.756 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation PADDING was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:25.756 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation PITCH was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:25.756 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation ZOOM was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:26.743 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation CENTER was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:26.743 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation BEARING was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:26.743 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation PADDING was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:26.743 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation PITCH was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:26.743 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation ZOOM was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:27.746 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation CENTER was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:27.746 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation BEARING was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:27.746 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation PADDING was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:27.746 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation PITCH was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:27.747 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation ZOOM was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:31.746 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation CENTER was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:31.746 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation BEARING was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:31.746 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation PADDING was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:31.746 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation PITCH was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:31.746 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation ZOOM was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:34.759 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation CENTER was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:34.759 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation BEARING was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:34.759 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation PADDING was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:34.759 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation PITCH was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:34.759 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation ZOOM was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:35.763 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation CENTER was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:35.763 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation BEARING was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:35.763 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation PADDING was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:35.763 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation PITCH was not registered and will not run. Register it with registerAnimation() method.
2022-09-02 12:00:35.763 21459-21848/? W/Mapbox: [maps-android\Mbgl-CameraAnimator]: Animation ZOOM was not registered and will not run. Register it with registerAnimation() method.
```
But it does not seem to be related to the actual hickup animation as it comes all the time.

**Comment by kiryldz** [comment]
Thanks for sharing this, I fixed several issues based on your findings please checkout latest snapshot `10.8.0-beta.1-MAPSAND-395-introduce-runtime-option-to-run-animators-on-non-main-thread-dd43c7cb2-SNAPSHOT`.

Side note: if you're creating and operating with any animator sets in your code directly - e.g. created:
```kotlin
val bearing = createBearingAnimator(...)
val center =  createCenterAnimator(...)
registerAnimators(bearing, center)
AnimatorSet().apply {
      playSequentially(*cameraAnimators.toTypedArray())
      start()
}
```
you __must__ also wrap `start()` call properly:
```kotlin
AnimationThreadController.postOnAnimatorThread { start() }
```

If you're using high-level animators (like `easeTo`) or medium-level (like `playAnimatorsTogether`) or even low-level (like `createBearingAnimator(...).start()`) - SDK will do that for you, no action required from your end. But due to Android SDK implementation if you're creating animator sets in your code base and start them - you need to wrap `start()` call on your end as well.

If you won't do that and use background thread - registering animators actually will happen on background thread while start will be triggered from main thread and that actually means race condition and 99% that registering will actually not be yet finished before starting and animation will not run.

**Comment by SamuelBrucksch** [comment]
We only use easeTo, so i guess it should be fine. So i guess we don't need to cancel either?

**Comment by SamuelBrucksch** [comment]
BTW dd43c7cb2 seems to work fine for easeTo. Thanks. So I will do some more testing and then try to get feedback from some Xiaomi users.

**Comment by kiryldz** [comment]
> We only use easeTo, so i guess it should be fine. So i guess we don't need to cancel either?

Cancelling high-level animations was fixed (hopefully) with last snapshot.

**Comment by SamuelBrucksch** [comment]
Out dev with the Xiaomi phone confirmed that it worked in an emulated drive. We also have some users testing that now on real trips. Will provide more feedback once we collected some data.

**Comment by SamuelBrucksch** [comment]
So we got some feedback that the map seems to move, no feedback from very long drives unfortunately, but at least on some shorter drives it worked.

**Comment by SamuelBrucksch** [comment]
Thanks for fixing this.

@kiryldz @kmadsen FYI we just got feedback, that on Xiaomi with Android 11 this workaround does not seem to work, however as the problem was introduced with Android 12, we added a check for xiaomi && >= android 12. You might want to apply that to your android auto implementation or samples if not having it already.
